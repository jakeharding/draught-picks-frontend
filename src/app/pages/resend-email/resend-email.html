<!--
  Generated template for the ResendEmailPage page.

  See http://ionicframework.com/docs/components/#navigation for more info on
  Ionic pages and navigation.
-->
<ion-header>
  <draught-header></draught-header>
</ion-header>

<ion-content>
  <ion-row>
    <ion-col class="ion-text-center"
             size="12" sizeSm="10" offsetSm="1" sizeMd="6" offsetMd="3" sizeLg="6" offsetLg="3" sizeXl="4" offsetXl="4">
      <form [formGroup]="emailForm">
        <h1>Resend Email</h1>
        <br>
        <p>Please enter the email address below:</p>
        <ion-item>
          <ion-label class="ion-text-left" position="floating">Email</ion-label>
          <ion-input type="text" formControlName="email"></ion-input>
        </ion-item>
      </form>
      <ion-button
          (click)="sendConfirmEmail()"
          margin-top="20px"
          [disabled]="emailForm.invalid">
        Submit
      </ion-button>
    </ion-col>
  </ion-row>

</ion-content>
