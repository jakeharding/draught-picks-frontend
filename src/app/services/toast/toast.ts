import { Injectable } from '@angular/core';
import { ToastController } from '@ionic/angular';

/*
  Generated class for the ToastProvider provider.

  See https://angular.io/guide/dependency-injection for more info on providers
  and Angular DI.
*/
@Injectable({
  providedIn: 'root',
})
export class ToastProvider {
  static readonly defaultErrorMsg = 'We are having trouble connecting at the moment. ' +
  'Please check your network connection and try again.';
  private readonly errorToastClass = 'error-toast';
  private readonly successToastClass = 'success-toast';

  constructor(private toastCtrl: ToastController) {}

  errorToast(message: string = ToastProvider.defaultErrorMsg): void {
    this.createToast(message, this.errorToastClass);
  }

  successToast(message: string): void {
    this.createToast(message, this.successToastClass);
  }

  async createToast(message: string,
                    cssClass: string,
                    duration: number = 3000,
                    position: 'top' | 'bottom' | 'middle' = 'top'
              ) {
    const toast = await this.toastCtrl.create({ message, cssClass, duration, position });
    await toast.present();
  }

}
