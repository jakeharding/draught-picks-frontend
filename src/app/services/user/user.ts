import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { environment as Env } from '../../../environments/environment';
import PageResponse from '../../models/PageResponse';
import User from '../../models/User';
import AuthResponse from '../../models/AuthResponse';
import BeerRating from '../../models/BeerRating';
import Beer from '../../models/Beer';

/**
 * Generated class for the UserProvider provider.
 *
 * See https://angular.io/guide/dependency-injection for more info on providers
 * and Angular DI.
 */
@Injectable({
  providedIn: 'root',
})
export class UserProvider {
  private readonly url: string;
  private readonly resendConfirmEmailUrl: string;
  private readonly confirmEmailUrl: string;
  private readonly passwordResetUrl: string;

  constructor(public http: HttpClient) {
    this.url = `${Env.REST_API_ROOT}/users`;
    this.resendConfirmEmailUrl = `${this.url}/resend-confirm-email`;
    this.confirmEmailUrl = `${this.url}/confirm-email`;
    this.passwordResetUrl = `${this.url}/reset-email`;
  }

  create(userData: AuthResponse | Beer | BeerRating | User) {
    return this.http.post(this.url, userData).toPromise();
  }

  retrieve() {
    return this.http.get(this.url).toPromise().then(({ results }: PageResponse<User>) => {
      return results[0];
    });
  }

  update(user: User): Promise<any> {
    return this.http.put(`${this.url}/${user.uuid}`, user).toPromise();
  }

  resendConfirmEmail({email}): Promise<object> {
    return this.http.post(this.resendConfirmEmailUrl, {email}).toPromise();
  }

  confirmEmail({key}): Promise<any> {
    return this.http.put(this.confirmEmailUrl, {confirm_key: key}).toPromise();
  }

  sendPasswordResetEmail(emailAddress: string): Promise<any> {
    return this.http.post(this.passwordResetUrl, {email: emailAddress}).toPromise();
  }
}
