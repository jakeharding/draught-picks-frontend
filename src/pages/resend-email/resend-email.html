<!--
  Generated template for the ResendEmailPage page.

  See http://ionicframework.com/docs/components/#navigation for more info on
  Ionic pages and navigation.
-->
<draught-header></draught-header>

<ion-content>
  <ion-row>
    <ion-col class="below-header" text-center col-12 col-sm-10 offset-sm-1 col-md-6 offset-md-3 col-lg-6 offset-lg-3 col-xl-4 offset-xl-4>
      <form [formGroupName]="emailForm">
        <ion-label text-left floating>Please enter the email address to resend the confirmation email below:</ion-label>
        <ion-input type="text" formControlName="email"></ion-input>
      </form>
    </ion-col>
  </ion-row>

</ion-content>
