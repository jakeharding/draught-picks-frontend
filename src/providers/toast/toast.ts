import { Injectable } from '@angular/core';
import { ToastController } from 'ionic-angular';

/*
  Generated class for the ToastProvider provider.

  See https://angular.io/guide/dependency-injection for more info on providers
  and Angular DI.
*/
@Injectable()
export class ToastProvider {
  private readonly errorToastClass = 'error-toast';
  private readonly successToastClass = 'success-toast';

  constructor(private toastCtrl: ToastController) {}

  errorToast(message: string): void {
    this.createToast(message, this.errorToastClass);
  }

  successToast(message: string): void {
    this.createToast(message, this.successToastClass);
  }

  createToast(message: string,
              cssClass:string,
              duration: number = 3000,
              position: string = 'top'
              ) {
    this.toastCtrl.create({ message, cssClass, duration, position }).present();
  }

}
