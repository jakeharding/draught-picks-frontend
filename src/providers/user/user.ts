import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import Env from "../../env";
import PageResponse from "../../models/PageResponse";
import User from "../../models/User";

/**
 * Generated class for the UserProvider provider.
 *
 * See https://angular.io/guide/dependency-injection for more info on providers
 * and Angular DI.
**/
@Injectable()
export class UserProvider {
  url: string;
  resendConfirmEmailUrl: string;

  constructor(public http: HttpClient) {
    this.url = `${Env.REST_API_ROOT}/users`;
    this.resendConfirmEmailUrl = `${this.url}/resend-confirm-email`;
  }

  create(userData: any) {
    return this.http.post(this.url, userData).toPromise();
  }

  retrieve() {
    return this.http.get(this.url).toPromise().then(({ results }: PageResponse<User>) => {
      return results[0];
    });
  }

  update(user: User): Promise<any> {
    return this.http.put(`${this.url}/${user.uuid}`, user).toPromise();
  }

  resendConfirmEmail({email}): Promise<Object> {
    return this.http.post(this.resendConfirmEmailUrl, {email}).toPromise();
  }
}
