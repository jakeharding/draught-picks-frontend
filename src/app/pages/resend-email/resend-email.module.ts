import { NgModule } from '@angular/core';
import { IonicModule } from '@ionic/angular';
import { ResendEmailPage } from './resend-email';
import { ComponentsModule } from '../../components/components.module';
import { RouterModule } from '@angular/router';
import { ReactiveFormsModule } from '@angular/forms';

@NgModule({
  declarations: [
    ResendEmailPage,
  ],
  imports: [
    IonicModule,
    RouterModule.forChild([{path: '', component: ResendEmailPage}]),
    ComponentsModule,
    ReactiveFormsModule,
  ],
})
export class ResendEmailPageModule {}
