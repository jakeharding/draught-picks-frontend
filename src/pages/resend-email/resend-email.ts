import { Component } from '@angular/core';
import { IonicPage, NavController, NavParams } from 'ionic-angular';
import { FormBuilder, FormGroup, Validators } from "@angular/forms";
import { BasePage } from "../BasePage";
import { UserProvider } from "../../providers/user/user";
import { ToastProvider } from "../../providers/toast/toast";

/**
 * Generated class for the ResendEmailPage page.
 *
 * See https://ionicframework.com/docs/components/#navigation for more info on
 * Ionic pages and navigation.
 */

@IonicPage()
@Component({
  selector: 'page-resend-email',
  templateUrl: 'resend-email.html',
})
export class ResendEmailPage extends BasePage {

  emailForm: FormGroup;

  constructor(public formBuilder: FormBuilder,
              public navCtrl: NavController,
              public navParams: NavParams,
              private toastProvider: ToastProvider,
              private userProvider: UserProvider) {
    super('resend-email');
    this.emailForm = this.formBuilder.group({
      email: ['', [Validators.required, Validators.email]]
    });
  }

  sendConfirmEmail() {
    this.userProvider.resendConfirmEmail(this.emailForm.value).then(() => {
      this.toastProvider.successToast(
        "Email has been sent. Please check you spam folder and visit this page again if you don't receive it.")
    }, () => {
      this.toastProvider.errorToast("Unable to send the email. Please check your connection and try again.")
    });
  }

}
