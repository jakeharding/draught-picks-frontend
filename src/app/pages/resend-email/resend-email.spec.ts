import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { ResendEmailPage } from './resend-email';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { NavController, NavParams } from '@ionic/angular';
import { FormBuilder, FormGroupDirective } from '@angular/forms';
import { ToastProvider } from '../../services/toast/toast';
import { UserProvider } from '../../services/user/user';

/**
 * resend-email.spec.ts
 *
 * Created by jake
 * Created on 2018-11-20
 *
 * Test ResendEmailPage.
 */

describe('ResendEmailPage', () => {
  let fixture: ComponentFixture<ResendEmailPage>;
  let component: ResendEmailPage;

  const mockUserProvider = {
    resendConfirmEmail: jest.fn(() => Promise.resolve())
  };

  const mockToastProvider = {
    successToast: jest.fn(),
    errorToast: jest.fn()
  };

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
      declarations: [ResendEmailPage, FormGroupDirective],
      providers: [
        { provide: NavController, useValue: {}},
        { provide: NavParams, useValue: {}},
        { provide: ToastProvider, useValue: mockToastProvider},
        { provide: UserProvider, useValue: mockUserProvider},
        FormBuilder
      ]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(ResendEmailPage);
    component = fixture.debugElement.componentInstance;
  });

  it('should initialize the ResendEmailPage', () => {
    expect(component).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });

  test('sendConfirmEmail successToast is called on promise resolve', async () => {
    await component.sendConfirmEmail();
    expect(mockUserProvider.resendConfirmEmail).toHaveBeenCalledTimes(1);
    expect(mockToastProvider.successToast).toHaveBeenCalledTimes(1);
  });

  test('sendConfirmEmail successToast is called on promise reject', async () => {
    mockUserProvider.resendConfirmEmail.mockReturnValue(Promise.reject('error'));
    await component.sendConfirmEmail();
    expect(mockUserProvider.resendConfirmEmail).toHaveBeenCalledTimes(1);
    expect(mockToastProvider.errorToast).toHaveBeenCalledTimes(1);
  });
});
