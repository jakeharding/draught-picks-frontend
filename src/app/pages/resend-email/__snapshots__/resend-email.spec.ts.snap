// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResendEmailPage should initialize the ResendEmailPage 1`] = `
<page-resend-email
  emailForm={[Function FormGroup]}
  formBuilder={[Function FormBuilder]}
  navCtrl={[Function Object]}
  toastProvider={[Function Object]}
  userProvider={[Function Object]}
>
  <ion-header>
    <draught-header />
  </ion-header><ion-content>
    <ion-row>
      <ion-col
        class="ion-text-center"
        offsetlg="3"
        offsetmd="3"
        offsetsm="1"
        offsetxl="4"
        size="12"
        sizelg="6"
        sizemd="6"
        sizesm="10"
        sizexl="4"
      >
        <form>
          <h1>
            Resend Email
          </h1>
          <br />
          <p>
            Please enter the email address below:
          </p>
          <ion-item>
            <ion-label
              class="ion-text-left"
              position="floating"
            >
              Email
            </ion-label>
            <ion-input
              formcontrolname="email"
              type="text"
            />
          </ion-item>
        </form>
        <ion-button
          margin-top="20px"
        >
           Submit 
        </ion-button>
      </ion-col>
    </ion-row>
  </ion-content>
</page-resend-email>
`;
